### DISCUSSION TITLE
What is the meaning / source of `Merged`, `Contained`, `Follows` in magit-log?

### DISCUSSION THREAD

**Discussion by kuanyui** [magit/magit#5105] [Support] [answered]
These two fields confuse me for a long time...

![Screenshot_20240308_153048](https://github.com/magit/magit/assets/1370070/41762bc5-21d6-4c48-93cf-378c622f4dd1)

How on Earth to understand the meaning of `Merged` and `Contained` in magit-log? It's not in the output of `git show COMMIT_ID`, also not found in Magit's official manual... Still cannot understand its behaviors after doing some experiments...

**Comment by tarsius** [comment, marked as answer]
From the `git-branch(1)` manpage:

>       With --contains, shows only the branches that contain the named commit
>       (in other words, the branches whose tip commits are descendants of the
>       named commit), --no-contains inverts it. With --merged, only branches
>       merged into the named commit (i.e. the branches whose tip commits are
>       reachable from the named commit) will be listed. With --no-merged only
>       branches not merged into the named commit will be listed. If the
>       <commit> argument is missing it defaults to HEAD (i.e. the tip of the
>       current branch).

**Comment by kuanyui** [comment]
Still quite incomprehensible and very difficult to memorize...

Curious that maybe adding tooltips on the texts of `Merged` `Contained` via text properties ( https://www.gnu.org/software/emacs/manual/html_node/elisp/Special-Properties.html ) may help user to understand?  @tarsius 

>- Contained: The branches that contain this commit. (i.e. the branches whose tip commits are descendants of this commit)
>- Merged: The branches merged into this commit (i.e. the branches whose tip commits are reachable from this commit) 

Actually, after comparing these descriptions with the real-world git repo, I still don't fully understand what `Merged` is, nor sure if the above quoted descriptions are correct...

**Comment by kuanyui** [comment]
And... May I know which command can get the `Follows` tag(s), or the keyword to the manpage?

![Screenshot_20240403_185041](https://github.com/magit/magit/assets/1370070/5cfd41f4-8723-4936-b948-fb2156f2e7c3)

**Reply by tarsius** [reply]
The tag and count are determined using this function:

```elisp
(defun magit-get-current-tag (&optional rev with-distance)
  "Return the closest tag reachable from REV.

If optional REV is nil, then default to `HEAD'.
If optional WITH-DISTANCE is non-nil then return (TAG COMMITS),
if it is `dirty' return (TAG COMMIT DIRTY). COMMITS is the number
of commits in `HEAD' but not in TAG and DIRTY is t if there are
uncommitted changes, nil otherwise."
  (and-let* ((str (magit-git-str "describe" "--long" "--tags"
                                 (and (eq with-distance 'dirty) "--dirty")
                                 rev)))
    (save-match-data
      (string-match
       "\\(.+\\)-\\(?:0[0-9]*\\|\\([0-9]+\\)\\)-g[0-9a-z]+\\(-dirty\\)?$" str)
      (if with-distance
          `(,(match-string 1 str)
            ,(string-to-number (or (match-string 2 str) "0"))
            ,@(and (match-string 3 str) (list t)))
        (match-string 1 str)))))
```

Note that this only works properly in repositories that do not contain any recent merges that go "in the wrong direction". A merge is in the wrong direction if someone works on `master` and then merges `origin/master` into that (their personal) `master`. This is wrong because this suggests that their `master`, which they have not shared with anyone at that point, was the "official version" all along, and that `origin/master` is a "feature branch", which is being merged now for the first time. They do that all the time in Emacs's repository, which is why you will get wildly incorrect results in that repository.

**Comment by kuanyui** [comment]
Finally, with help of LLM, I made (yet another) a dirty hack for personal usage. I reword and add tooltips for them.

I can't help but do this, because the meaning of the word still confuse me every time I see this page. I'm just unable to memorize **"Who on Earth is the object of the verb `Merged`?", "Who merged who?, who contains who?"** after using Magit & git over 10 years.

This is what I preferred / expected:

<img width="494" height="145" alt="スクリーンショット_20260515_133503" src="https://github.com/user-attachments/assets/ca4f90b9-d8e7-45ce-994f-3fbc9b74aa3a" />
<img width="494" height="162" alt="スクリーンショット_20260515_133517" src="https://github.com/user-attachments/assets/4d3bfd6e-3361-45a6-94bd-a99f5302467a" />

Code:
```lisp
(with-eval-after-load 'magit-diff
  ;; Dirty hack for `magit-insert-revision-headers' to replace the ambiguous wording
  ;; (`Merged', `Contained') with clearer labels and add tooltips.
  (defvar my-magit-related-refs-rename
    '(("Merged"
       :to "Merged into"
       :help "[git branch --merged <this commit>]\nThis commit is already merged into these branches:")
      ("Contained"
       :to "Ancestor of"
       :help "[git branch --contains <this commit>]\nThis commit is a parent of these branches:"))
    "Rewrite rules for Magit related-refs section headings.
       Each entry is (ORIGINAL :to NEW :help TOOLTIP).")

  (define-advice magit--insert-related-refs
      (:filter-args (args) my-rename-related-refs-heading)
    "Rename the Merged / Contained section headings."
    (if-let* ((orig (nth 2 args))
              (rule (cdr (assoc orig my-magit-related-refs-rename))))
        (let ((new  (plist-get rule :to))
              (tip  (plist-get rule :help)))
          (list (nth 0 args)
                (nth 1 args)
                (propertize new 'help-echo tip)
                (nth 3 args)))
      args))

  ;; Override built-in `magit--insert-related-refs' in `magit-diff.el'
  (defun magit--insert-related-refs (rev arg title remote)
    (when-let ((refs (magit-list-related-branches arg rev (and remote "-a"))))
      (insert title ":" (make-string (- 12 (length title)) ?\s))   ;; <--- Upstream was 10
      (dolist (branch refs)
        (if (<= (+ (current-column) 1 (length branch))
                (window-width))
            (insert ?\s)
          (insert ?\n (make-string 14 ?\s)))    ;; <--- Upstream was 12
        (magit-insert-section (branch branch)
          (insert (propertize branch 'font-lock-face
                              (if (string-prefix-p "remotes/" branch)
                                  'magit-branch-remote
                                'magit-branch-local)))))
      (insert ?\n)))
  )

```
